---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [23]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [24]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [25]:
def degree_distribution(G):
    degrees = G.degree()
    degree_values = sorted(set(degrees.values()))
    histogram = [list(degrees.values()).count(i)/float(nx.number_of_nodes( G)) for i in degree_values]
    return histogram

In [26]:
def graph_identification():
    
    # Your Code Here
    methods = []
    for G in P1_Graphs:
        clustering = nx.average_clustering(G)
        shortest_path = nx.average_shortest_path_length(G)
        degree_hist = degree_distribution(G)
        if len(degree_hist)>10:
            methods.append('PA')
        elif clustering < 0.1:
            methods.append('SW_H')
        else:
            methods.append('SW_L')
    return methods
graph_identification()

['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [27]:
G = nx.read_gpickle('email_prediction.txt')

print(nx.info(G))

Name: 
Type: Graph
Number of nodes: 1005
Number of edges: 16706
Average degree:  33.2458


### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [28]:
df = pd.DataFrame(index=G.nodes())
df['department'] = pd.Series(nx.get_node_attributes(G, 'Department'))
df['clustering'] = pd.Series(nx.clustering(G))
df['degree'] = pd.Series(G.degree())
df['salary'] = pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))
df['degree_centrality'] = pd.Series(nx.degree_centrality(G))
df['closeness'] = pd.Series(nx.closeness_centrality(G, normalized=True))
df['betweeness'] = pd.Series(nx.betweenness_centrality(G, normalized=True))
df['pr'] = pd.Series(nx.pagerank(G))

In [29]:
df.head()

department  clustering  degree  salary  degree_centrality  closeness  \
0           1    0.276423      44     0.0           0.043825   0.421991   
1           1    0.265306      52     NaN           0.051793   0.422360   
2          21    0.297803      95     NaN           0.094622   0.461490   
3          21    0.384910      71     1.0           0.070717   0.441663   
4          21    0.318691      96     1.0           0.095618   0.462152   

   betweeness        pr  
0    0.001124  0.001224  
1    0.001195  0.001426  
2    0.006570  0.002605  
3    0.001654  0.001833  
4    0.005547  0.002526

In [30]:
df_train = df[~pd.isnull(df['salary'])]
df_test = df[pd.isnull(df['salary'])]

In [31]:
features = ['clustering', 'degree', 'degree_centrality', 'closeness', 'betweeness', 'pr']

In [32]:
X_train=df_train[features]
y_train=df_train['salary']
X_test=df_test[features]

In [33]:
def salary_predictions():
    
    # Your Code Here
    from sklearn.svm import SVC
    from sklearn.neural_network import MLPClassifier
    from sklearn.preprocessing import MinMaxScaler
    scaler = MinMaxScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    clf = MLPClassifier(hidden_layer_sizes = [10, 5], alpha = 5,
                       random_state = 0, solver='lbfgs', verbose=0)
    clf.fit(X_train_scaled, y_train)
    test_proba = clf.predict_proba(X_test_scaled)[:, 1]
    ans=pd.Series(test_proba,X_test.index)
    return ans

In [34]:
salary_predictions()

1       0.128439
2       0.586261
5       0.979859
8       0.133702
14      0.304625
18      0.208766
27      0.269586
30      0.336144
31      0.169191
34      0.143513
37      0.096221
40      0.195036
45      0.162267
54      0.249954
55      0.252905
60      0.176908
62      0.992842
65      0.551284
77      0.078478
79      0.132108
97      0.099812
101     0.057517
103     0.214342
108     0.122930
113     0.454489
122     0.044350
141     0.396296
142     0.863703
144     0.072049
145     0.419966
          ...   
913     0.045734
914     0.078520
915     0.002087
918     0.144304
923     0.035175
926     0.090079
931     0.055969
934     0.013933
939     0.011147
944     0.002353
945     0.039608
947     0.162715
950     0.098173
951     0.038313
953     0.027345
959     0.002922
962     0.002440
963     0.139886
968     0.110279
969     0.104672
974     0.095061
984     0.012668
987     0.110375
989     0.113364
991     0.116944
992     0.003956
994     0.003077
996     0.0024

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [35]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [36]:
for node in G.nodes():
        G.node[node]['community'] = G.node[node]['Department']

In [37]:
future_connections['preferential attachment'] = [i[2] for i in nx.preferential_attachment(G, future_connections.index)]
future_connections['Common emp'] = future_connections.index.map(lambda emp: len(list(nx.common_neighbors(G, emp[0], emp[1]))))
future_connections['jaccard'] = [i[2] for i in nx.jaccard_coefficient(G, future_connections.index)]
future_connections['rs alloc'] = [i[2] for i in nx.resource_allocation_index(G, future_connections.index)]
future_connections['adar'] = [i[2] for i in nx.adamic_adar_index(G, future_connections.index)]

In [38]:
future_connections['hopcroft']=[i[2] for i in nx.cn_soundarajan_hopcroft(G, future_connections.index)]

In [39]:
df1=future_connections.copy()

In [40]:
df1.head()

Future Connection  preferential attachment  Common emp   jaccard  \
(6, 840)                  0.0                     2070           9  0.073770   
(4, 197)                  0.0                     3552           2  0.015504   
(620, 979)                0.0                       28           0  0.000000   
(519, 872)                0.0                      299           2  0.060606   
(382, 423)                0.0                      205           0  0.000000   

            rs alloc      adar  hopcroft  
(6, 840)    0.136721  2.110314         9  
(4, 197)    0.008437  0.363528         2  
(620, 979)  0.000000  0.000000         0  
(519, 872)  0.039726  0.507553         2  
(382, 423)  0.000000  0.000000         0

In [41]:
df1.columns

Index(['Future Connection', 'preferential attachment', 'Common emp', 'jaccard',
       'rs alloc', 'adar', 'hopcroft'],
      dtype='object')

In [42]:
features=['preferential attachment', 'Common emp', 'jaccard','rs alloc', 'adar', 'hopcroft']

In [43]:
df_train1=df1[~pd.isnull(df1['Future Connection'])]
df_test1=df1[pd.isnull(df1['Future Connection'])]

In [44]:
X_train1=df_train1[features]
y_train1=df_train1['Future Connection']
X_test1=df_test1[features]

In [45]:
def new_connections_predictions():
    
    # Your Code Here
    from sklearn.svm import SVC
    from sklearn.neural_network import MLPClassifier
    from sklearn.preprocessing import MinMaxScaler
    scaler = MinMaxScaler()
    X_train_scaled = scaler.fit_transform(X_train1)
    X_test_scaled = scaler.transform(X_test1)
    clf = MLPClassifier(hidden_layer_sizes = [10, 5], alpha = 5,
                       random_state = 0, solver='lbfgs', verbose=0)
    clf.fit(X_train_scaled, y_train1)
    test_proba = clf.predict_proba(X_test_scaled)[:, 1]
    ans=pd.Series(test_proba,X_test1.index)
    return ans

In [49]:
new_connections_predictions()

(107, 348)    0.026711
(542, 751)    0.010518
(20, 426)     0.586460
(50, 989)     0.010572
(942, 986)    0.010630
(324, 857)    0.010564
(13, 710)     0.157547
(19, 271)     0.134784
(319, 878)    0.010591
(659, 707)    0.010524
(49, 843)     0.010607
(208, 893)    0.010537
(377, 469)    0.008422
(405, 999)    0.017790
(129, 740)    0.016576
(292, 618)    0.020991
(239, 689)    0.010594
(359, 373)    0.009263
(53, 523)     0.232484
(276, 984)    0.010605
(202, 997)    0.010616
(604, 619)    0.061416
(270, 911)    0.010593
(261, 481)    0.070565
(200, 450)    0.999906
(213, 634)    0.010501
(644, 735)    0.048282
(346, 553)    0.010367
(521, 738)    0.010108
(422, 953)    0.017680
                ...   
(672, 848)    0.010593
(28, 127)     0.932317
(202, 661)    0.010344
(54, 195)     0.999807
(295, 864)    0.010556
(814, 936)    0.010510
(839, 874)    0.010630
(139, 843)    0.010561
(461, 544)    0.009928
(68, 487)     0.009896
(622, 932)    0.010543
(504, 936)    0.016642
(479, 528) 